distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Dialect.enso

```diff
@@ -310,68 +317,112 @@ type SQLServer_Dialect
-                    # If the function expects more arguments, we pass the metadata as the last argument.
-                    case result of
-                        _ : Function -> result metadata
-                        _ -> result
-            null_checks_result = if kind == "IS_NULL" then [] else null_checks
-
-            
-            pair expr_result null_checks_result
-        query : Query -> pair (base_gen.generate_sub_query self query) []
-
-    ## PRIVATE
-    generate_column_for_select self base_gen expr:(SQL_Expression | Order_Descriptor | Query) name:Text -> SQL_Builder =
-        expr_null_checks_pair = self.generate_expression base_gen expr
-        base_expr = expr_null_checks_pair.first
-        built_expr = case expr of
-            SQL_Expression.Operation _ _ _ ->
-                null_check = if expr_null_checks_pair.second.length == 0 then "" else
-                    SQL_Builder.code "WHEN " ++ _generate_null_check_sql_builder expr_null_checks_pair.second ++ " THEN NULL "
-                SQL_Builder.code "CASE " ++ null_check ++ "WHEN " ++ base_expr ++ " THEN 1 ELSE 0 END"
-            _ -> base_expr
-        built_expr ++ Base_Generator.alias self name
+    generate_expression self base_gen expr:(SQL_Expression | Order_Descriptor | Query) for_select:Boolean -> SQL_Builder =
+        if for_select then _generate_column self base_gen expr else
+            _generate_expression self base_gen expr generate_as_bool=True . first
+
+## Returns a pair of a SQL_Builder for the given expression and a vector of columns
+   that have been used in the expression and need to be checked for nulls.
+   SQL Server needs special handling commpared to ther databases as it does not have a
+   boolean data type.
+   This means that you can write
+       SELECT * FROM MyTable WHERE [Column1] > [Column2]
+   but you cannot write
+       SELECT [Column1] > [Column2] FROM MyTable
+   to write the second query you need to write
+       SELECT CASE WHEN [Column1] IS NULL OR [Column2] IS NULL WHEN [Column1] > [Column2] THEN 1 ELSE 0 END FROM MyTable
+   The below function collects all of the fields which are needed to be checked for nulls returning them in a vector
+   as the second element of the pair.
+   The first element of the pair is the SQL_Builder for the expression.
+private _generate_expression dialect base_gen expr generate_as_bool = case expr of
+    SQL_Expression.Column _ _ ->
+        wrapped_name = base_gen.generate_expression dialect expr
+        typed_result = case generate_as_bool of 
+            False -> wrapped_name
+            True -> wrapped_name ++ " = 1"
+        pair typed_result [wrapped_name]
+    SQL_Expression.Constant value -> 
+        wrapped = case value of
+            Nothing -> SQL_Builder.code "NULL"
+            _ -> SQL_Builder.interpolation value
+        pair wrapped [wrapped]
+    SQL_Expression.Literal value -> 
+        wrapped = SQL_Builder.code value
+        pair wrapped [wrapped]
```

Comment:
Please note that this may also require some handling for boolean conversions, because we have a few places in our _common_ codebase that contains `SQL_Expression.Literal "TRUE"` and similar.

I'm ok with not handling it in this PR if it's not necessary to make `filter` feature work. But I'd probably stick a TODO somewhere here. Maybe we need to refactor these to avoid relying on literal and instead introduce 'fake' operations `SQL_Expression.Operation "TRUE" []` that will generate the true/false literals but also will allow to control the return types using the same facility as other operations.

Reply:
I don't generally like untested code. But I like TODOs even less so added the conversions :)